### PR TITLE
Add etcd based Mesos master detector plugin.

### PR DESCRIPTION
**Pull request by gradywang**
Currently, the etcd-based master contender/detector module is developing in Mesos community, refer to MESOS-1806 for the details. This patch will add a etcd based mesos master detector plugin，then when we deploy the kubernetes cluster on a high-availability model Mesos cluster, it does not need to run another zookeeper cluster just for Mesos.

### CODE REVIEW

**Comment by gradywang** [comment]
@jdef Cloud you help to review this PR for me?

**Comment by jdef** [comment]
will do. I'm on PTO for the next several business days, returning on 8/4.
Will take a look then

**Comment by gradywang** [comment]
@jdef cloud you help to take a look at this patch when you have time, thanks!

File: detector/etcd/client.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+// Etcd watch event actions
+const (
+	EtcdUpdate = "update"
+	EtcdSet    = "set"
+	EtcdCAS    = "compareAndSwap"
+)
+
+// Configurations for etcd client
+type ETCDClientCfg struct {
```
Comment (jdef) [inline detector/etcd/client.go]: nit: be consistent w/ naming things. For example `EtcdUpdate` vs. `ETCDClientconfig` - pick one style and stick with it.

File: detector/etcd/client.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+	etcd "github.com/coreos/etcd/client"
+	log "github.com/golang/glog"
+	"golang.org/x/net/context"
+	"sync"
+	"time"
+)
+
+// Etcd watch event actions
+const (
+	EtcdUpdate = "update"
```
Comment (jdef) [inline detector/etcd/client.go]: nit: why are these public? and why do they start with `Etcd` (because as is, they stutter)? possible alternatives: `actionUpdate`, `eventUpdate`, `watchEventUpdate` .. or some such

File: detector/etcd/client.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package etcd
+
+import (
+	etcd "github.com/coreos/etcd/client"
```
Comment (jdef) [inline detector/etcd/client.go]: what version of etcd? only asking because I thought there were significant API changes between releases. if not, nevermind

File: detector/etcd/doc.go [hunk 4]
```diff
@@ -0,0 +1,3 @@
+// Etcd-based mesos-master leaderhip detection.
+// Implements support for optional detector.AllMasters interface.
```
Comment (jdef) [inline detector/etcd/doc.go]: nope, this doesn't appear to be true

File: detector/etcd/detect.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+		f.OnMasterChanged(nil)
+		return
+	}
+
+	masterInfo := &mesos.MasterInfo{}
+
+	if err := json.Unmarshal(pair.value, masterInfo); err != nil {
+		log.Errorln("Unmarshal the master information %a failed, due to: ", string(pair.value), err.Error())
+		md.leaderNode = ""
+		f.OnMasterChanged(nil)
```
Comment (jdef) [inline detector/etcd/detect.go]: logPanic

File: detector/etcd/detect.go [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+		case <-time.After(md.minDetectorCyclePeriod - elapsed): // noop
+		}
+	}
+}
+
+// notifyMasterChanged notifies the scheduler when Mesos master changed
+func (md *MasterDetector) notifyMasterChanged(pair *kvPair, f detector.MasterChanged) {
+	if pair == nil {
+		md.leaderNode = ""
+		f.OnMasterChanged(nil)
```
Comment (jdef) [inline detector/etcd/detect.go]: logPanic

File: detector/etcd/detect.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+func (md *MasterDetector) detect(f detector.MasterChanged) {
+	log.V(3).Infoln("detecting children at", md.masterDetectPath)
+
+	for {
+		isDetectorClosed := false
+		select {
+		case <-md.Done():
+			isDetectorClosed = true
+			break
```
Comment (jdef) [inline detector/etcd/detect.go]: redundant

File: detector/etcd/detect.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+		for {
+			started := time.Now()
+			select {
+			case pairNode := <-md.client.MasterChangedChan():
+				md.notifyMasterChanged(pairNode, f)
+			case err := <-md.client.MasterDetectErrChan():
+				log.V(1).Infoln("Watch ended with error, master lost; error was:", err)
+				isWatchFailed = true
+				md.notifyMasterChanged(nil, f)
+				break
```
Comment (jdef) [inline detector/etcd/detect.go]: redundant

File: detector/etcd/detect.go [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+		md := di.(*MasterDetector)
+		old := md.client
+		md.client = usedClient
+		md.cancel = usedClient.Stop
+		return UsedEtcdClient(old)
+	}
+}
+
+// detectedCycle defines the master detected cycle
+func (md *MasterDetector) detectedCycle(started time.Time) {
```
Comment (jdef) [inline detector/etcd/detect.go]: nit: detectedCycle .. might be better named `maybeWaitAfterDetected` or something that describes what it actually does.

File: detector/etcd/detect.go [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+func MinCyclePeriod(d time.Duration) detector.Option {
+	return func(di interface{}) detector.Option {
+		md := di.(*MasterDetector)
+		old := md.minDetectorCyclePeriod
+		md.minDetectorCyclePeriod = d
+		return MinCyclePeriod(old)
+	}
+}
+
+func UsedEtcdClient(usedClient ETCDInterface) detector.Option {
```
Comment (jdef) [inline detector/etcd/detect.go]: naming: UsedEtcdClient doesn't read very well. maybe something like `WithEtcdClient`?

Comment (jdef) [inline detector/etcd/detect.go]: or better .. `WithClient` (less stutter)

File: detector/etcd/detect.go [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+const (
+	defaultMinDetectorCyclePeriod      = 1 * time.Second
+	defaultEtcdHeaderTimeoutPerRequest = 3 * time.Second
+)
+
+type (
+	ETCDInterface interface {
+		Watch(string, uint64)
+		Get(string) (*kvPair, error)
+		Put(string, string) error
```
Comment (jdef) [inline detector/etcd/detect.go]: why is this needed?

File: detector/etcd/client.go [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+		outgoing:  make(chan *kvPair),
+		client:    etcd.NewKeysAPI(c),
+		done:      make(chan struct{}),
+		stopped:   false,
+		ctx:       nil,
+		cancel:    nil,
+	}
+
+	go func() {
+		defer close(s.etcdError)
```
Comment (jdef) [inline detector/etcd/client.go]: this is racy if `Stop` is ever called while a watch is in progress. Possible that these chan's will close and then a write to either of them will panic

**Comment by gradywang** [comment]
@jdef Thank you very much for your review comments, I will go through them and fix them asap, thank you again.

File: detector/etcd/detect.go (hunk 6 above)
Comment (gradywang) [inline detector/etcd/detect.go]: Fixed.

File: detector/etcd/detect.go (hunk 5 above)
Comment (gradywang) [inline detector/etcd/detect.go]: Fixed.

**Comment by gradywang** [comment]
@jdef, travis checks failed due to I used `*http.Request.Cancle()` interface in this PR, this interface does not be supported in the golang versions which < 1.5, do we have to support those earlier golang versions, can we change the supported smallest version to 1.5 in mesos-go?

In addition, could you help to review my fixes for your comments? Thanks!

**Comment by jdef** [comment]
xref https://issues.apache.org/jira/browse/MESOS-1806

i'll hopefully take another in-depth look at this soon. it's probably good to go. sorry it's taken me so long to get back to this.

**Comment by xiang90** [comment]
@jdef 

if the motivation is to share etcd between mesos and k8s, it should be targeted at v3 api not v2. this pr is targeting at v2.

**Comment by jdef** [comment]
is this feature still of interest to anyone?

**Comment by jdef** [comment]
This look abandoned, will close. Please re-open if you think this is closed in error.
